Fix OAuth redirect URI format to align with Microsoft's URL standards

bluedog13 opened the pull request:
## OAuth Redirect URI Format Not Aligned with Microsoft's URL Standards

### Problem
VS Code's OAuth implementation generates redirect URIs without trailing slashes, while Microsoft's own OAuth 2.0 documentation and standards-compliant identity providers (including Entra ID) require trailing slashes in redirect URIs per RFC specifications.

This creates a mismatch where:
- **VS Code generates**: `http://127.0.0.1:3000`
- **IDP expects**: `http://127.0.0.1:3000/`

### Root Cause
Microsoft Entra ID (and other compliant IDPs) automatically append trailing slashes to redirect URIs during OAuth flows, following Microsoft's documented standards. However, VS Code's implementation returns URLs without trailing slashes, causing URI mismatch validation failures.

### Recent Change Context
A June 29th PR that modified `loopbackServer.ts` removed the trailing slash, introducing this standards compliance issue. This change made VS Code generate non-compliant redirect URIs that don't align with Microsoft's own OAuth 2.0 format requirements.

### Solution
This PR restores trailing slashes in two key locations:

1. **`loopbackServer.ts`** (line 105): Runtime redirect URI generation
2. **`oauth.ts`** (lines 739-746): Dynamic client registration redirect URIs

**Changes:**
- `http://127.0.0.1:${port}` → `http://127.0.0.1:${port}/`
- `http://localhost` → `http://localhost/`
- `http://127.0.0.1` → `http://127.0.0.1/`
- `http://localhost:${port}` → `http://localhost:${port}/`

### Impact
- Resolves OAuth authentication failures with "redirect_uri_mismatch" errors
- Ensures compatibility with standards-compliant OAuth providers
- Aligns VS Code with Microsoft's documented OAuth 2.0 URL format requirements

### Standards Reference
This change ensures VS Code follows Microsoft's own OAuth 2.0 redirect URI format requirements, improving compatibility with standards-compliant identity providers.

Fixes #260425

gautam-goudar committed:
Fix OAuth redirect URI validation by adding trailing slash

This change adds a trailing slash to the redirect URI returned by the
loopbackServer's redirectUri getter to match the expected format for
OAuth client registration.

The issue occurs because Microsoft Entra ID automatically appends
a trailing slash to redirect URIs without a path segment, but VS Code
was returning URLs without the trailing slash, causing validation
failures during OAuth authentication.

Fixes #260425

Commit:
Add trailing slashes to OAuth redirect URIs in dynamic registration

Complements the loopbackServer.ts fix by ensuring redirect URIs used in
OAuth dynamic client registration also include trailing slashes for
complete standards compliance with Microsoft's OAuth 2.0 URL format.

Updated redirect URIs:
- http://localhost -> http://localhost/
- http://127.0.0.1 -> http://127.0.0.1/
- http://localhost:{port} -> http://localhost:{port}/
- http://127.0.0.1:{port} -> http://127.0.0.1:{port}/

bluedog13 commented on src/vs/workbench/api/node/loopbackServer.ts:
```diff
@@ -102,7 +102,7 @@ export class LoopbackAuthServer implements ILoopbackServer {
 		if (this._port === undefined) {
 			throw new Error('Server is not started yet');
 		}
-		return `http://127.0.0.1:${this._port}`;
```
Have updated the PR with changes to src/vs/base/common/oauth.ts

TylerLeonhardt reviewed (dismissed):
LGTM, just need you to agree to the contributor agreement and we can get this in!

bluedog13 commented:
@microsoft-github-policy-service agree

TylerLeonhardt commented:
@bluedog13 can you update the current one please

Commit:
Fix OAuth redirect URI test expectations to match implementation

The fetchDynamicRegistration function correctly includes trailing slashes
on redirect URIs per the implementation. This updates the test expectations
to match the actual behavior, fixing the test failure.

The implementation in oauth.ts includes trailing slashes on localhost URIs:
- 'http://localhost/' (with trailing slash)
- 'http://127.0.0.1/' (with trailing slash)
- 'http://localhost:/' (with trailing slash)
- 'http://127.0.0.1:/' (with trailing slash)

This test was expecting URIs without trailing slashes, causing the assertion
to fail. The fix aligns the test with the correct implementation behavior.

TylerLeonhardt commented:
Thanks for contributing @bluedog13!
